pkg/frontend/querymiddleware/astmapper/sharding.go

```diff
@@ -456,13 +456,27 @@ func shardMatrixSelector(curshard, shards int, selector *parser.MatrixSelector)
 	return nil, fmt.Errorf("invalid selector type: %T", selector.VectorSelector)
 }
 
-// isSubquery returns true if the given function call expression is a subquery.
+// isSubquery returns true if the given function call expression is a subquery,
+// or a subquery wrapped by parenthesis.
 func isSubquery(n *parser.Call) bool {
-	if len(n.Args) == 0 {
+	for _, arg := range n.Args {
+		if ok := isSubqueryVisitFn(arg); ok {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isSubqueryVisitFn(expr parser.Expr) bool {
```

Comment:
I think this one is the one that should be called `isSubquery`, and the one above should be `isSubqueryAggregationCall` or similar.

I didn't like the name of the previous (because it lies, it says is subquery but actually checks a call's arguments), and now it's even clearer to me.

Reply:
I can't see why it can't be used standalone, but I won't oppose to that.